src/function/gds/louvain.cpp

```diff
@@ -0,0 +1,483 @@
+                        maxIndex > sc) { // Swap protection
+                        maxIndex = sc;
+                    }
+                    max = maxIndex;
+                }
+                state.targetCommunities[nodeId] = max;
+            } else {
+                state.targetCommunities[nodeId] = UNASSIGNED_COMM;
+            }
+
+            auto currCommId = state.currCommunities[nodeId];
+            auto targetCommId = state.targetCommunities[nodeId];
+            if (targetCommId != currCommId && targetCommId != UNASSIGNED_COMM) {
+                state.commInfosNext[targetCommId].addDegree(state.nodeWeightedDegrees[nodeId]);
+                state.commInfosNext[targetCommId].addSize(1);
+                state.commInfosNext[currCommId].addDegree(-state.nodeWeightedDegrees[nodeId]);
+                state.commInfosNext[currCommId].addSize(-1);
+            }
+            clusterLocalMap.clear();
+            counter.clear();
+        }
+    }
+
+    std::unique_ptr<InMemVertexCompute> copy() override {
+        return std::make_unique<RunIterationVC>(state);
+    }
+
+private:
+    PhaseState& state;
+};
+
+class NewModularityVC : public InMemVertexCompute {
+public:
+    static std::atomic<double> sumIntra;
+    static std::atomic<double> sumTotal;
+
+    explicit NewModularityVC(PhaseState& state) : state{state} {}
+
+    void vertexCompute(offset_t startOffset, offset_t endOffset) override {
+        double sumIntraLocal = 0.0;
+        double sumTotalLocal = 0.0;
+        for (auto nodeId = startOffset; nodeId < endOffset; ++nodeId) {
+            sumIntraLocal += state.clusterWeightInternal[nodeId];
+            sumTotalLocal += state.commInfos[nodeId].degree * state.commInfos[nodeId].degree;
+        }
+        sumIntra.fetch_add(sumIntraLocal);
+        sumTotal.fetch_add(sumTotalLocal);
+    }
+
+    std::unique_ptr<InMemVertexCompute> copy() override {
+        return std::make_unique<NewModularityVC>(state);
+    }
+
+private:
+    PhaseState& state;
+};
+std::atomic<double> NewModularityVC::sumIntra{0.0};
+std::atomic<double> NewModularityVC::sumTotal{0.0};
+
+class UpdateCommInfosVC : public InMemVertexCompute {
+public:
+    static std::atomic<double> sumIntra;
+    static std::atomic<double> sumTotal;
+
+    explicit UpdateCommInfosVC(PhaseState& state) : state{state} {}
+
+    void vertexCompute(offset_t startOffset, offset_t endOffset) override {
+        for (auto nodeId = startOffset; nodeId < endOffset; ++nodeId) {
+            state.commInfos[nodeId].size += state.commInfosNext[nodeId].getSize();
+            state.commInfos[nodeId].degree += state.commInfosNext[nodeId].getDegree();
+        }
+    }
+
+    std::unique_ptr<InMemVertexCompute> copy() override {
+        return std::make_unique<UpdateCommInfosVC>(state);
+    }
+
+private:
+    PhaseState& state;
+};
+
+class WriteResultsVC : public GDSResultVertexCompute {
```

Comment:
Is it possible to refactor and re-use some other ResultsWriter? What is needed so that we do not keep writing our own results writers.

Reply:
I don't see this addressed.

Reply:
Will work on this in a future PR.